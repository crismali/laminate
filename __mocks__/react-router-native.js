/* eslint-env jest */
const { Switch, Route, matchPath } = require.requireActual('react-router-native')
const withRouter = component => component

export { withRouter, Switch, Route, matchPath }
