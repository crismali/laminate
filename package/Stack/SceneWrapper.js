import { Component } from 'react'
import { Animated } from 'react-native'
import PropTypes from 'prop-types'
import { renderChildrenComponentWithContext } from '../renderChildrenComponentWithContext'

export class SceneWrapper extends Component {
  static propTypes = {
    animationValue: PropTypes.instanceOf(Animated.Value).isRequired,
    children: PropTypes.any.isRequired,
    index: PropTypes.number.isRequired
  }

  static childContextTypes = {
    index: PropTypes.number.isRequired,
    interpolateAnimation: PropTypes.func.isRequired,
    setAnimationValue: PropTypes.func.isRequired
  }

  getChildContext () {
    return {
      index: this.props.index,
      interpolateAnimation: this.interpolateAnimation,
      setAnimationValue: this.setAnimationValue
    }
  }

  setAnimationValue = (value) => {
    const { animationValue, index } = this.props
    animationValue.setValue(index + value)
  }

  interpolateAnimation = ({ inputRange, outputRange }) => {
    const { animationValue, index } = this.props
    return animationValue.interpolate({
      inputRange: inputRange.map(value => value + index),
      outputRange
    })
  }

  render () {
    return this.props.children
  }
}

export const WithStackAnimation = renderChildrenComponentWithContext('WithStackAnimation', SceneWrapper.childContextTypes)
