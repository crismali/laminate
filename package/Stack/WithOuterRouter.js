import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { WrapRouter } from './WrapRouter'

export class OuterHistory extends Component {
  static propTypes = {
    children: PropTypes.any.isRequired
  }

  static contextTypes = {
    outerRouter: PropTypes.object.isRequired
  }

  render () {
    return <WrapRouter router={this.context.outerRouter}>
      {this.props.children}
    </WrapRouter>
  }
}
