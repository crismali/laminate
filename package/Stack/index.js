export { Stack } from './Stack'
export { WithOuterRouter } from './WithOuterRouter'
