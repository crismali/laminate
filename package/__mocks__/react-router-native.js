/* eslint-env jest */
import React from 'react'

const { Switch, Route, matchPath, StaticRouter } = require.requireActual('react-router-native')
const withRouter = component => component
const MemoryRouter = (props) => <StaticRouter context={{}} {...props} />

export { withRouter, Switch, Route, matchPath, MemoryRouter, StaticRouter }
