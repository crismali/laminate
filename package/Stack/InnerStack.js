import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { View, StyleSheet, Text } from 'react-native'
import { Animation } from '../Animation'
import { SceneWrapper } from './SceneWrapper'

const styles = StyleSheet.create({
  container: {
    width: '100%',
    height: '100%'
  }
})

const allButLast = (collection) => collection.slice(0, collection.length - 1)

export class InnerStack extends Component {
  static propTypes = {
    ...Animation.childContextTypes,
    children: PropTypes.any.isRequired,
    innerRouter: PropTypes.object.isRequired,
    outerRouter: PropTypes.object.isRequired
  }

  static childContextTypes = {
    outerRouter: PropTypes.object.isRequired
  }

  state = {
    scenes: [this.props.children]
  }

  getChildContext () {
    return {
      outerRouter: this.props.outerRouter
    }
  }

  componentDidMount () {
    this.props.animation.play()
  }

  componentWillReceiveProps (nextProps) {
    const { innerRouter: newInnerRouter } = nextProps
    const scenesLength = this.scenes().length
    const { index: historyIndex } = newInnerRouter.history

    switch (newInnerRouter.history.action) {
      case 'REPLACE':
        this.handleReplace(nextProps); break
      case 'PUSH':
        if (scenesLength !== historyIndex + 1) {
          this.handlePush(nextProps)
        }
        break
      case 'POP':
        this.handlePop(nextProps)
        break
    }
  }

  scenes = () => this.state.scenes

  handlePop = (nextProps) => {
    const scenes = this.scenes().slice(0, nextProps.innerRouter.history.index + 1)
    this.props.animation.rewind({
      toValue: scenes.length,
      callback: () => this.setState({ scenes })
    })
  }

  handlePush = (nextProps) => {
    const scenes = [...this.scenes(), nextProps.children]
    this.setState({ scenes })
    setTimeout(() => nextProps.animation.play({ toValue: scenes.length }), 10)
  }

  handleReplace = (nextProps) => {
    const scenes = [...allButLast(this.scenes()), nextProps.children]
    this.setState({ scenes })
  }

  render () {
    return <View style={styles.container}>
      {this.scenes().map((scene, index) => (
        <SceneWrapper
          key={index}
          index={index}
          animationValue={this.props.animation.value}
        >
          {scene}
        </SceneWrapper>
      ))}
    </View>
  }
}
