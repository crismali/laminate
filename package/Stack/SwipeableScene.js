import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Route } from 'react-router-native'
import { SwipeCatcher } from '../SwipeCatcher'
import { WithStackAnimation } from './SceneWrapper'

class InnerSwipeableScene extends Component {
  static defaultProps = {
    completeThreshold: 0.5
  }

  static propTypes = {
    children: PropTypes.any.isRequired,
    completeThreshold: PropTypes.number,
    history: PropTypes.object.isRequired,
    location: PropTypes.object.isRequired,
    setAnimationValue: PropTypes.func.isRequired
  }

  handleMove = ({ percentComplete, velocity }) => {
    this.props.setAnimationValue(1 - percentComplete)
  }

  handleGestureRelease = ({ percentComplete, velocity }) => {
    const { completeThreshold, history, location } = this.props

    if (percentComplete > completeThreshold) {
      history.goBack()
    } else {
      history.replace(location)
    }
  }

  handleGestureTerminate = ({ percentComplete, velocity }) => {
    const { history, location } = this.props
    history.replace(location)
  }

  render () {
    return <SwipeCatcher
      {...this.props}
      onMove={this.handleMove}
      onGestureRelease={this.handleGestureRelease}
      onGestureTerminate={this.handleGestureTerminate}
    />
  }
}

export class SwipeableScene extends Component {
  render () {
    return <WithStackAnimation>
      {({ setAnimationValue }) => (
        <Route>
          {({ history, location }) => (
            <InnerSwipeableScene
              {...this.props}
              setAnimationValue={setAnimationValue}
              history={history}
              location={location}
            />
          )}
        </Route>
      )}
    </WithStackAnimation>
  }
}
