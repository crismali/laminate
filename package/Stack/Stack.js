import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Route, Switch } from 'react-router-native'
import { Animation } from '../Animation'
import { InnerStack } from './InnerStack'

export class Stack extends Component {
  static defaultProps = {
    initialLocation: '/'
  }

  static propTypes = {
    children: PropTypes.any.isRequired,
    easing: PropTypes.any,
    duration: PropTypes.number,
    initialLocation: PropTypes.any.isRequired
  }

  render () {
    return <Route>
      {(router) => (
        <Animation easing={this.props.easing} duration={this.props.duration}>
          {animation => (
            <InnerStack animation={animation} router={router}>
              <Switch location={router.location}>
                {this.props.children}
              </Switch>
            </InnerStack>
          )}
        </Animation>
      )}
    </Route>
  }
}
