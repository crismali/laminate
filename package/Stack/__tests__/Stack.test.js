/* eslint-env jest */
import React from 'react'
import { mount } from 'enzyme'
import { StaticRouter, Route } from 'react-router-native'
import { Stack } from '../Stack'

const Child = () => <div>a child</div>

it('renders properly', () => {
  const subject = mount(<StaticRouter location='/bar' context={{}}>
    <Stack initialLocation='/foo' easing='easing' duration={123}>
      <Route component={Child} path='/foo' />
    </Stack>
  </StaticRouter>)
  expect(subject).toMatchSnapshot()
})
