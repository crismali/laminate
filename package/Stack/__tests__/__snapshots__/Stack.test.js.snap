// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<StaticRouter
  basename=""
  context={Object {}}
  location="/bar"
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/bar",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Stack
      duration={123}
      easing="easing"
      initialLocation="/foo"
    >
      <Route>
        <Animation
          duration={123}
          easing="easing"
        >
          <InnerStack
            animation={
              Object {
                "play": [Function],
                "rewind": [Function],
                "value": Value {
                  "value": 0,
                },
              }
            }
            router={
              Object {
                "history": Object {
                  "action": "POP",
                  "block": [Function],
                  "createHref": [Function],
                  "go": [Function],
                  "goBack": [Function],
                  "goForward": [Function],
                  "listen": [Function],
                  "location": Object {
                    "hash": "",
                    "pathname": "/bar",
                    "search": "",
                  },
                  "push": [Function],
                  "replace": [Function],
                },
                "location": Object {
                  "hash": "",
                  "pathname": "/bar",
                  "search": "",
                },
                "match": Object {
                  "isExact": false,
                  "params": Object {},
                  "path": "/",
                  "url": "/",
                },
                "staticContext": Object {},
              }
            }
          >
            <SceneWrapper
              animationValue={
                Value {
                  "value": 0,
                }
              }
              index={0}
              key="0"
            >
              <Switch
                location={
                  Object {
                    "hash": "",
                    "pathname": "/bar",
                    "search": "",
                  }
                }
              />
            </SceneWrapper>
          </InnerStack>
        </Animation>
      </Route>
    </Stack>
  </Router>
</StaticRouter>
`;
